src/test/HST/Util/SubstTests.hs

```diff
@@ -61,124 +56,310 @@ currentFrontendSyntax = return
-            [S.PVar S.NoSrcSpan x, S.PVar S.NoSrcSpan x_0]
-            (S.Tuple S.NoSrcSpan S.Boxed [S.var x, S.var x_0, S.var y])
-      e' <- currentFrontendSyntax $ applySubst subst e
-      expected <- currentFrontendSyntax
-        $ S.Lambda S.NoSrcSpan [S.PVar S.NoSrcSpan x_0, S.PVar S.NoSrcSpan x_1]
-        (S.Tuple S.NoSrcSpan S.Boxed [S.var x_0, S.var x_1, S.var x])
-      setExpectation (e' `shouldBe` expected)
+      let subst    = singleSubst (S.unQual y) (S.var x)
+          e        = ["\\x   x_0 -> (x, x_0, y)"]
+          expected = ["\\x_0 x_1 -> (x_0, x_1, x)"]
+      in shouldSubstituteTo subst e expected
+    it "does not substitute bound variables in let expressions"
+      $ runTest
+      $
+      -- σ    = { x ↦ z, y ↦ z }
+      -- e    = let x = z in x y
+      -- σ(e) = let x = z in x z
+      let subst
+            = substFromList [(S.unQual x, S.var z), (S.unQual y, S.var z)]
+          e        = ["let x = z in x y"]
+          expected = ["let x = z in x z"]
+      in shouldSubstituteTo subst e expected
+    it "renames variables that are no longer bound"
+      $ runTest
+      $
+      -- σ    = { y ↦ z }
+      -- e    = let x = let y = z in y
+      --        in x z
+      -- σ(e) = let x = let y = z in y
+      --        in x y
+      let subst    = singleSubst (S.unQual y) (S.var z)
+          e        = ["let x = let y = z in y", "in x y"]
+          expected = ["let x = let y = z in y", "in x z"]
+      in shouldSubstituteTo subst e expected
+    it "does not substitute bound variables in nested let expressions"
+      $ runTest
+      $
+      -- σ    = { x ↦ z, y ↦ z }
+      -- e    = let x = y
+      --        in let y = z in x y
+      -- σ(e) = let x = z
+      --        in let y = z in x y
+      let subst
+            = substFromList [(S.unQual x, S.var z), (S.unQual y, S.var z)]
+          e        = ["let x = y", "in let y = z in x y"]
+          expected = ["let x = z", "in let y = z in x y"]
+      in shouldSubstituteTo subst e expected
+    it "does not use more variable names than needed"
+      $ runTest
+      $
+      -- σ    = { y ↦ x }
+      -- e    = let x = y
+      --        in let x = y in x
+      -- σ(e) = let x_0 = x
+      --        in let x_0 = x in x_0
+      let subst    = singleSubst (S.unQual y) (S.var x)
+          e        = ["let x = y", "in let x = y in x"]
+          expected = ["let x_0 = x", "in let x_0 = x in x_0"]
+      in shouldSubstituteTo subst e expected
+    it "avoids capture by renaming variables in let expressions"
+      $ runTest
+      $
+      -- σ    = { y ↦ x, z ↦ x }
+      -- e    = let x   = y
+      --            x_0 = z
+      --        in  x_1
+      -- σ(e) = let x_0 = x
+      --            x_2 = x
+      --        in  x_1
+      let subst
+            = substFromList [(S.unQual y, S.var x), (S.unQual z, S.var x)]
+          e        = ["let x   = y", "    x_0 = z", "in  x_1"]
+          expected = ["let x_0 = x", "    x_2 = x", "in  x_1"]
```

Comment:
The following is easier to read.

```suggestion
          e        = ["let { x   = y; x_0 = z } in x_1"]
          expected = ["let { x_0 = x; x_2 = x } in x_1"]
```

I think we should avoid multi-line expressions that are written without actual line breaks in general. Thus, I'd remove `unlines` from `parseTestExpression` and use `{}`-syntax where possible.